Added Teleport Hook, Updated HookNotify Plugin

joshi07 opened the pull request:
i added a Hook for the teleport. 

Plugins may or may not deny any teleportation.

Updated HookNotify plugin with it.

as requested in  #1538

NiLSPACE commented:
I don't think this is right. You can also teleport entities, so I think you should either rename this to OnPlayerTeleport, or make it accept all entities. I vote for accept all entities ;)

joshi07 commented:
you're right, i will look over this to accept all entities. :) 

why did the Travis CI thingy fail ? it just compiled fine on my Visual express ?!

i hoped my first contribution would go through in first try xD *dreaming*

NiLSPACE commented on src/Entities/Player.cpp:
```diff
@@ -1274,13 +1274,19 @@ unsigned int cPlayer::AwardAchievement(const eStatistic a_Ach)
 
-	m_World->BroadcastTeleportEntity(*this, GetClientHandle());
-	m_ClientHandle->SendPlayerMoveLook();
+	if (!cRoot::Get()->GetPluginManager()->CallHookPlayerTeleport(*this, m_LastPos, GetPosition())) // if the Plugin doesn't denie the teleport
```
You need two spaces before you start a comment
```
MyFunction)  // My comment
```

That's why the build failed

joshi07 commented:
Thank you very much, i will think on that next time :)

joshi07 commented:
is it possible to call the hook from multiple funktions? 

like from this places ?

cEntity::TeleportToCoords
and
cPlayer::TeleportToCoords 

so the hook would accept all entities ?

NiLSPACE commented:
Yes, you can call it from wherever you want.

NiLSPACE commented:
You do have to change the parameters in the hook itself though ;)

joshi07 commented:
how can i add changes from my rebo to this PR ?

joshi07 commented:
Yay it passed :) if you want to hook for player teleports, you have to use the IsPlayer funktion to find out if the Entity is a player or some other stuff.

madmaxoft commented on src/Entities/Entity.cpp:
```diff
@@ -1632,8 +1632,11 @@ void cEntity::TeleportToEntity(cEntity & a_Entity)
 {
-	SetPosition(a_PosX, a_PosY, a_PosZ);
-	m_World->BroadcastTeleportEntity(*this);
+	if (!cRoot::Get()->GetPluginManager()->CallHookEntityTeleport(*this, m_LastPos, GetPosition()))  // if the Plugin doesn't denie the teleport
```
`deny` is the proper spelling. Also, it might be a good idea to put the comment above the condition, so that the line is not that long. I'd write it like this:
```cpp
// If plugins agree with the teleport, set the new position:
if (!cRoot::Get()->...)
```

madmaxoft replied:
Does the `GetPosition()` return the new position already? I don't think so.

madmaxoft commented on src/Entities/Player.cpp:
```diff
@@ -1274,13 +1274,19 @@ unsigned int cPlayer::AwardAchievement(const eStatistic a_Ach)
 
-	m_World->BroadcastTeleportEntity(*this, GetClientHandle());
-	m_ClientHandle->SendPlayerMoveLook();
+	if (!cRoot::Get()->GetPluginManager()->CallHookEntityTeleport(*this, m_LastPos, GetPosition()))  // if the Plugin doesn't denie the teleport
```
Same as in cEntity - put the comment above and check that `GetPosition()` is the right thing to call.

joshi07 commented:
thank you, i will look into that.

joshi07 commented:
now this works, it returns the new position, so the plugin can decide to allow the teleport.

madmaxoft commented:
Could you rebase onto current master, and squash those "fix spaces" commits?

joshi07 commented:
i can and will tomorrow! And with "return", i didn't mean the hook in C++  but the funktion in LUA :)

madmaxoft commented:
The Lua function doesn't return anything except for the `bool` either, does it?

joshi07 commented:
You're right, i need to work on my writing, so my writing matches my thougts!

It was late and i was tired yesterday.

What is it called, if the funktion call comes with arguments for the function? (Thats what i meant with return for the LUA) 

Does it "pass" the arguments? I've never talked in english about programming.

madmaxoft commented:
I'd use "function takes the new position as an argument", but I'm not a native English speaker either, so it needn't be the best one.
I think that all those four lines of comments are unneeded, only the first line, "Ask plugins if they allow the teleport" should be kept, remove the rest.

worktycho commented:
I am a native speaker and in my experience you'd use "function takes the new position as an argument". Generally I use takes for describing the interface, and passes for describing the action of calling the function.

joshi07 commented:
im sorry to keep asking stupid question. But how can i rebase my fork and add my changes again? do i have to delete my fork and refork? i tried some stuff but i coudn't resolve it. Im pretty new to Github.

joshi07 commented:
i think i f*cked up :o how can i clean up that mess ? maybe its best to delete my fork and readd my changes ?so i get a clean hystory ?

madmaxoft commented:
Normally I work with command-line git, so here's what I'd do:
```
git remote add upstream https://github.com/mc-server/MCServer.git
git fetch --all
git rebase --interactive upstream/master
```
At this point, git will open your favorite text editor to ask you to edit a file that has a list of all the commits, along with instructions; you can select an action for each commit. Save the file and exit, git will continue with the rebase. If there are any conflicts, you will need to fix them, commit and continue with the rebase.

joshi07 commented:
for every commit which is not mine i get this error with untracked files.

am i doing something wrong? I corrected the conflicts by changing the conflicting files content and used 
`git add *filefapth*` 
to add the file.
then i put
`git rebase --continue`
After that when my commits are done and he tries to replay the other commits (none of mine) this error pops up.
```
rebase in progress; onto d4b505d
You are currently rebasing branch 'master' on 'd4b505d'.

Untracked files:
        sfmtcpoy.cqq.txt

nothing added to commit but untracked files present
The previous cherry-pick is now empty, possibly due to conflict resolution.
If you wish to commit it anyway, use:

    git commit --allow-empty

Otherwise, please use 'git reset'
Could not apply cfce882e2f003979728da7fa77c2c5df43b26cb8... Fixed door placement.
```
i don't know what to do there

madmaxoft commented:
I guess in this situation the easiest would be to store your changes elsewhere, reset the git repo back to upstream/master and re-do the changes.
